### ISSUE TITLE
Parallel demo + tests

### ISSUE THREAD

**Issue by CodyCBakerPhD** [catalystneuro/tqdm_publisher#31]
Here is a simple script that approximates the generic parallel structure we'd use with this package

```python
"""Demo of parallel tqdm."""
import time
from typing import List, Tuple
from concurrent.futures import ProcessPoolExecutor

import tqdm


def _run_sleep_in_subprocess(args: Tuple[int, int]):
    """The operation to run on each subprocess."""
    repeat = args[0]
    iteration_index = args[1]

    iterable = range(repeat)
    sub_progress_bar = tqdm.tqdm(
        iterable=iterable, position=iteration_index + 1, desc=f"Progress on iteration {iteration_index}", leave=False
    )
    for _ in sub_progress_bar:
        time.sleep(2.5)


def run(number_of_jobs: int, repeats: List[int]):
    """The main (outer) iteration run on the central managing process."""
    total_iterations = len(repeats)
    with ProcessPoolExecutor(max_workers=number_of_jobs) as executor:
        job_map = executor.map(
            _run_sleep_in_subprocess, [(repeat, iteration_index) for iteration_index, repeat in enumerate(repeats)]
        )
        main_progress_bar = tqdm.tqdm(
            iterable=job_map,
            position=0,  # Total progress bar at top position
            total=total_iterations,
            desc="Total progress",
        )

        for _ in main_progress_bar:  # perform iteration to deploy jobs
            main_progress_bar.update()


if __name__ == "__main__":
    number_of_jobs = 3
    repeats = [4, 6, 8, 10, 4, 8, 20, 10, 5, 12, 5, 4, 5, 5, 5]

    run(number_of_jobs=number_of_jobs, repeats=repeats)
```

save this as a `.py` file and call from a terminal, it ought to look something like

https://github.com/catalystneuro/tqdm_publisher/assets/51133164/6f926f4f-dbb6-4996-b0cd-4cdc6bf4c230

Now, you'll notice a number of things there that aren't ideal with the way it renders in the terminal; there are no doubt other TQDM parameters that can optimize the appearance, but the point of this package I think is still to not overfit those aspects to the stdout of a terminal; the `n` counter of the outer TQDM bar was indeed actually getting called on each main iteration (deployment to subprocess) regardless of the terminals rendering, and I'm hoping by having a separate renderer with your subscription callbacks that would be easier to make it look better

**Comment by CodyCBakerPhD** [comment]
Also: I originally really wanted the ability to only have as many subprogress bars as workers to prevent the screen from getting too spammed over a large number of outer iterations; but after playing with it I do realize how tricky that is when each iteration takes a random amount of time; you can't cleanly pre-map each individual run with a specific process so you'd need a mechanism for tracking which window is available on the start of a job, but that could lead to race conditions, etc.

**Comment by garrettmflynn** [comment]
Alright, so I've refactored this to highlight the problem I've been having, which is independent of the way we map each process onto a global progress bar: 
```
"""Demo of parallel tqdm."""
from tqdm_publisher import TQDMPublisher
import time
from typing import List, Tuple
from concurrent.futures import ProcessPoolExecutor
import uuid

id_ref = uuid.uuid4()

def send_to_website(id, data):
    print('Send to website', id_ref, id, data)

class ParallelProgressHandler:
    def __init__(self):
        self.callbacks = {}
        self.started = False

    def _run_sleep_in_subprocess(self, args: Tuple[int, int]):
        """The operation to run on each subprocess."""
        repeat = args[0]
        iteration_index = args[1]
        id = args[2]

        iterable = range(repeat)
        sub_progress_bar = TQDMPublisher(
            iterable=iterable, position=iteration_index + 1, desc=f"Progress on iteration {iteration_index} ({id})", leave=False
        )

        # 1. Have the callbacks be the same as from the main process (DOES NOT WORK)
        # sub_progress_bar.subscribe(lambda data: self._run_callbacks(id, data))

        # 2. Have the function and all references be passed from the main process (DOES NOT WORK)
        # sub_progress_bar.subscribe(lambda data: send_to_website(id, data))

        # 3. Forward the updates to the main process (IF POSSIBLE, WOULD BE IDEAL)
        def to_main_process(data):
            print('To main process', id, data) # Add code to forward to the main process

        sub_progress_bar.subscribe(to_main_process)


        for _ in sub_progress_bar:
            time.sleep(.5)

    def _run_callbacks(self, id, data):
        for callback in self.callbacks.values():
            callback(id, data)


    def run(self, number_of_jobs: int, repeats: List[int]):
        """The main (outer) iteration run on the central managing process."""

        if (self.started):
            return
        
        self.started = True

        with ProcessPoolExecutor(max_workers=number_of_jobs) as executor:

            job_map = executor.map(
                self._run_sleep_in_subprocess, [(repeat, iteration_index, uuid.uuid4()) for iteration_index, repeat in enumerate(repeats)]
            )
            
            [_ for _ in job_map] # perform iteration to deploy jobs

            # 3. Run all updates through the user-defined callback
            def on_update_from_subprocess(data):
                self._run_callbacks(id, data)


    def subscribe(self, callback):
        id = uuid.uuid4()
        self.callbacks[id] = callback
        return id
    

if __name__ == "__main__":
    number_of_jobs = 3
    repeats = [4, 6, 8, 10, 4, 8, 20, 10, 5, 12, 5, 4, 5, 5, 5]

    def send_to_website(id, data):
        print('Send to website', id, data)

    parallel_progress_handler = ParallelProgressHandler()
    parallel_progress_handler.run(number_of_jobs=number_of_jobs, repeats=repeats)
    parallel_progress_handler.subscribe(send_to_website)
```

Generally, I have a single callback on the main process that I'd like to run whenever any of the subprocess `tqdm` updates occur. This would require me to (1) maintain a reference to a class object across processes, (2) call a function with all the same references in each process, or (3) forward the updates from each process to the main process. 

(1) and (2) appear fundamentally impossible in Python, unless I've missed something. And I'm not sure how to implement (3). Does this make my problem any clearer?

**Comment by CodyCBakerPhD** [comment]
What is the `data` in all your callables? I can't really understand what it's supposed to execute from looking at this

At the end of your script, shouldn't you subscribe before running? Calling `run` as it is should run the sleep operations before you ever define the `on_update_from_subprocess` or call `_run_callbacks`

**Comment by garrettmflynn** [comment]
`data` is just the `format_dict` forwarded from the tqdm update function.

You're correct about the placement of `on_update_from_subprocess`. However, there's currently no place to call this / subscribe this to for the desired effect (i.e. run every time one of the subprocesses has an update)—which is precisely the issue I'm trying to raise.

Updating in response to the `main_progress_bar` (which I've removed here) is trivial, since it's just tracking when each of the subprocesses completes on the main process.

**Comment by CodyCBakerPhD** [comment]
> data is just the format_dict forwarded from the tqdm update function.

OK would be helpful to rename it as such in the future to make that clearer; otherwise, `data` could be the thing being iterated, the state of the bar (`total`/`n`), or something external to TQDM itself



> However, there's currently no place to call this / subscribe this to for the desired effect (i.e. run every time one of the subprocesses has an update)—which is precisely the issue I'm trying to raise.

Sorry, the abstraction/nesting of your example is hard to read and harder to figure out exactly what you want to happen in what order

The callback you have here is 'send to website'; what is it sending? And is there a reason that spawned workers can't send it directly to the website?

I like the `TQDMPublisher` injection design because it's so flexible (just an arbitrary callable that can be run within the scope of a `tqdm.tqdm` instance and so have access to all internals), but that means that the communication workflow to whatever external subscriber needs to be designed a certain way

What I was imaging, correct me if nonsensical, was some external asynchronous server, possibly accessed via a port on local host, that our callbacks can send a request to containing information about a given task such as (`iteration_index`), current progress (`n`), and total progress (`total`)? In which case each subprocess only needs to know how to send those messages

If that's not the structure for whatever reason, there's also a way we can just do timed monitor updates on the main process (like every X seconds) that reads from shared memory that gets updated by `tqdm.tqdm.update` from each subprocess


> Updating in response to the main_progress_bar (which I've removed here) is trivial, since it's just tracking when each of the subprocesses completes on the main process.

Right, so we're talking about the other progress bars after the first; the ones whose update steps are only directly accessible to the namespace in the memory of the subprocess

**Comment by garrettmflynn** [comment]
> OK would be helpful to rename it as such in the future to make that clearer; otherwise, data could be the thing being iterated, the state of the bar (total/n), or something external to TQDM itself

Got it. I'll make sure to do that for our demo code.

> Right, so we're talking about the other progress bars after the first; the ones whose update steps are only directly accessible to the namespace in the memory of the subprocess

Exactly. Unfortunately, the demo code relies on a WebSocket connection that is detected on the main process, and these connections can only be initiated by a _client_ (e.g. the HTML code we have written). As such, we can't replicate that connection in each subprocess.

> What I was imaging, correct me if nonsensical, was some external asynchronous server, possibly accessed via a port on local host, that our callbacks can send a request to containing information about a given task such as (iteration_index), current progress (n), and total progress (total)? In which case each subprocess only needs to know how to send those messages

This is an interesting idea. This would work, as I could spawn an HTTP server that each process submits updates to. Though this would then require us to introduce such a server on all the dependents  that use subprocesses for their tqdm updates (`nwbinspector`, `dandi-cli`, etc.) to get this to work. 

If that's agreeable, I can move forward with this demo scheme!

> If that's not the structure for whatever reason, there's also a way we can just do timed monitor updates on the main process (like every X seconds) that reads from shared memory that gets updated by tqdm.tqdm.update from each subprocess

This is also fine and a bit more minimal, but removes the streamlined aspects of the original pub/sub design (since we're sampling arbitrarily, not getting notified of each update).

**Comment by CodyCBakerPhD** [comment]
> If that's agreeable, I can move forward with this demo scheme!

Let's at least make an additional demo to try it out and see what it looks like

**Comment by CodyCBakerPhD** [comment]
closed by #51
